### PR TITLE
fixed transaction addresses mapping, fixes #1971

### LINKED ISSUE #1971: parity trace_transaction showing internal calls incorrectly
the parity trace for http://etherscan.io/vmtrace?txhash=0x7725f11b932394f67bb07dcecff9bda0dd529a4a12a98be7cf418de8d36bb2d8&type=parity shows 2 internal calls

While, a similiar trace against geth shows no calls http://etherscan.io/vmtrace?txhash=0x7725f11b932394f67bb07dcecff9bda0dd529a4a12a98be7cf418de8d36bb2d8

The address is 0x2ab9f67a27f606272189b307052694d3a2b158ba is EOA (not a contract address) and shouldn't be generating any internal calls ?

### PR DESCRIPTION
**Pull request by debris**
changes:
- only transactions from canon chain will be stored in mapping TransactionHash -> TransactionAddress
- tests

### CODE REVIEW

**Comment by arkpar** [comment]
Receipts must be updated as well

**Comment by arkpar** [comment]
Ideally, any info on transactions that are removed from canon chain and never added back should be deleted.

File: ethcore/src/blockchain/blockchain.rs [hunk 1]
```diff
@@ -961,15 +961,45 @@ impl BlockChain {
...
+							block_hash: info.hash.clone(),
+							index: i
+						})
+					})
+					.collect()
+			},
+			BlockLocation::BranchBecomingCanonChain(ref data) => {
+				let addresses = data.enacted.iter()
+					.map(|hash| (hash, self.block_body(hash).unwrap()))
+					.flat_map(|(hash, bytes)| {
```
Comment (rphmeier) [inline ethcore/src/blockchain/blockchain.rs]: should be an expect and the `map` and `flat_map` here can probably be unified.

**Comment by debris** [comment]
> Receipts must be updated as well

Receipts are stored in hashmap, where the key is block hash (`HashMap<BlockHash, BlockReceipt>`), so they don't have to be updated.

> Ideally, any info on transactions that are removed from canon chain and never added back should be deleted. But if it make is to complicated, I'm fine with them sticking around.

It will make it slightly more complicated and I would prefer to do it in separate pr. The pr would be focus on removing all the unused data from db

File: ethcore/src/blockchain/blockchain.rs (hunk 1 above)
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: 👍

**Comment by arkpar** [comment]
Right, I was assuming the key for receipts is tx hash

**Pull request merged**